src/compiler/crystal/semantic/bindings.cr

```diff
@@ -1,7 +1,101 @@
 module Crystal
+  # Specialized container for ASTNodes to use for bindings tracking.
+  #
+  # The average number of elements in both dependencies and observers is below 2
+  # for ASTNodes. This struct inlines the first two elements saving up 4
+  # allocations per node (two arrays, with a header and buffer for each) but we
+  # need to pay a slight extra cost in memory upfront: a total of 6 pointers (48
+  # bytes) vs 2 pointers (16 bytes). The other downside is that since this is a
+  # struct, we need to be careful with mutation.
+  struct SmallNodeList
+    include Indexable(ASTNode)
```

Comment:
**thought:** It might be worth noting that this collection type implements a subset of `Array`'s features, but the missing bits are not relevant to the use case.
I'm also wondering if this actually needs to implement `Indexable`. `Enumerable` should actually be sufficient.
The only place were indexed access is needed is for a deconstruction `first, second = nodes` in `type_merge` and that's easy to refactor. This would reduce overall code by removing `unsafe_fetch`.

Reply:
Good point, thanks! I'll make the change.